Tests/SwiftGRPCNIOTests/GRPCChannelHandlerTests.swift

```diff
@@ -0,0 +1,196 @@
+import Foundation
+import XCTest
+import NIO
+import NIOHTTP1
+@testable import SwiftGRPCNIO
+
+func gRPCMessage(channel: EmbeddedChannel, compression: Bool = false, message: Data? = nil) -> ByteBuffer {
+  let messageLength = message?.count ?? 0
+  var buffer = channel.allocator.buffer(capacity: 5 + messageLength)
+  buffer.write(integer: Int8(compression ? 1 : 0))
+  buffer.write(integer: UInt32(messageLength))
+  if let bytes = message {
+    buffer.write(bytes: bytes)
+  }
+  return buffer
+}
+
+class GRPCChannelHandlerTests: GRPCChannelHandlerResponseCapturingTestCase {
+  func testUnimplementedMethodReturnsUnimplementedStatus() throws {
+    let responses = try waitForGRPCChannelHandlerResponses(count: 1) { channel in
+      let requestHead = HTTPRequestHead(version: .init(major: 2, minor: 0), method: .POST, uri: "unimplemented")
+      try channel.writeInbound(RawGRPCServerRequestPart.head(requestHead))
+    }
+
+    XCTAssertNoThrow(try extractStatus(responses[0])) { status in
+      XCTAssertEqual(status.code, .unimplemented)
+    }
+  }
+
+  func testImplementedMethodReturnsHeadersMessageAndStatus() throws {
+    let responses = try waitForGRPCChannelHandlerResponses(count: 3) { channel in
+      let requestHead = HTTPRequestHead(version: .init(major: 2, minor: 0), method: .POST, uri: "/echo.Echo/Get")
+      try channel.writeInbound(RawGRPCServerRequestPart.head(requestHead))
+
+      let request = Echo_EchoRequest.with { $0.text = "echo!" }
+      let requestData = try request.serializedData()
+      var buffer = channel.allocator.buffer(capacity: requestData.count)
+      buffer.write(bytes: requestData)
+      try channel.writeInbound(RawGRPCServerRequestPart.message(buffer))
+    }
+
+    XCTAssertNoThrow(try extractHeaders(responses[0]))
+    XCTAssertNoThrow(try extractMessage(responses[1]))
+    XCTAssertNoThrow(try extractStatus(responses[2])) { status in
+      XCTAssertEqual(status.code, .ok)
+    }
+  }
+
+  func testImplementedMethodReturnsStatusForBadlyFormedProto() throws {
+    let responses = try waitForGRPCChannelHandlerResponses(count: 2) { channel in
+      let requestHead = HTTPRequestHead(version: .init(major: 2, minor: 0), method: .POST, uri: "/echo.Echo/Get")
+      try channel.writeInbound(RawGRPCServerRequestPart.head(requestHead))
+
+      var buffer = channel.allocator.buffer(capacity: 3)
+      buffer.write(bytes: [1, 2, 3])
+      try channel.writeInbound(RawGRPCServerRequestPart.message(buffer))
+    }
+
+    XCTAssertNoThrow(try extractHeaders(responses[0]))
+    XCTAssertNoThrow(try extractStatus(responses[1])) { status in
+      let expectedStatus = GRPCStatus.requestProtoParseError
+      XCTAssertEqual(status.code, expectedStatus.code)
```

Comment:
Should we make `GRPCStatus` `Equatable` or `Hashable`, to simplify this?

Reply:
We can do that, however, we'd also be checking the trailing metadata. These tests collect the responses _before_ the `grpc-{status, message}` get added to the trailers so they should be fine.

Reply:
Makes sense, resolving this.